errors/errors.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors provides an error type for Go Cloud APIs.
+package errors
+
+import "fmt"
+
+// A Code describes the error's category. Programs should act upon an error's
+// code, not its message.
+type Code int
+
+const (
+	// The error could not be categorized.
+	Unknown Code = iota
+
+	// The resource was not found.
+	NotFound
+
+	// The resource exists, but it should not.
+	AlreadyExists
+
+	// The caller is not authorized to perform the operation.
+	PermissionDenied
+
+	// A value given to a Go Cloud API is incorrect.
+	InvalidArgument
+
+	// Something unexpected happened. Internal errors always indicate
+	// bugs in Go Cloud (or possibly the underlying provider).
+	Internal
+)
+
+var codeStrings = []string{
+	"Unknown",
+	"NotFound",
+	"AlreadyExists",
+	"PermissionDenied",
+	"InvalidArgument",
+	"Internal",
+}
+
+// An Error described a Go Cloud error.
+type Error struct {
+	Code Code
+	msg  string
+	err  error
+}
+
+func (e *Error) Error() string {
+	cs := "?badCode?"
+	if e.Code >= 0 && int(e.Code) < len(codeStrings) {
+		cs = codeStrings[e.Code]
+	}
+	return fmt.Sprintf("%s: %s", cs, e.msg)
+}
+
+// TODO(jba): use the new errors formatting code to print the underlying error.
+
+// Unwrap returns the error underlying the receiver.
+func (e *Error) Unwrap() error {
+	return e.err
+}
+
```

Comment:
IIUC, the proposed replacement for `blob.IsNotExist(err)` is `gerrors.ErrorCode(err) == gerrors.NotFound`, right?

That seems less discoverable to me, mostly because it's not obvious from the return type (error) that you should be looking at this other error package (gerrors).

Maybe we can just keep the existing `blob.IsNotFound` etc.? And/or add `blob.ErrorCode(err)` that returns the gerrors.ErrorCode?

Similarly, what's the proposed replacement for

```
var aerr aws.Error
if blob.ErrorAs(&aerr) {
  // use aerr.
}
```
?

I think something like:

```
rawerr := gerrors.Unwrap(err)
if rawerr != nil {
  if aerr, ok := rawerr.(aws.Error); ok {
    // Use aerr.
  }
}
```

Again, that seems rather unwieldy in comparison, and also doesn't match the `As` style used elsewhere. Maybe we should keep `blob.ErrorAs` and `driver.ErrorAs`...?

Reply:
We would document the error codes returned in all functions where it mattered, and we'd have examples. People using our APIs would figure out pretty quickly that we have our own errors package with its own type and codes. Especially since that's how both GCP and AWS do it: gRPC has a `Status`, Amazon has `aws.Error`.

I'm fine with leaving a legacy boolean function in blob, but let's rename it IsNotFound for consistency. I wouldn't add new ones to other packages, though, since they're redundant.

The replacement for `ErrorAs` is `errors.As`, the general `As` function that is currently in `golang.org/x/exp/errors` but will be in the stdlib `errors` in, we think,Go 1.13.

Reply:
* I think if we document a few error codes in specific functions, and add something to the package documentation pointing to our errors package, that should be OK.

* I'm fine with just removing the legacy boolean function I guess; we would keep it for backwards compatibility, so there's not much point in renaming it.

* Unless you have a strong objection let's keep `ErrorAs` in our concrete types (and in our conformance tests). For the driver interface, I'm less sure it's necessary to keep `ErrorAs`, but maybe we should for flexibility (e.g., one of the reasons for `As` was to allow providers to support multiple target types). The implementation for each driver can use `golang.org/x/exp/errors` or stay the same as it is now.

Reply:
> let's keep `ErrorAs` in our concrete types

I don't have a strong objection now, but when `errors.As` is in the standard library, it will be the idiomatic way to do this, and we should be in the vanguard in adopting it. So as long as we remove `ErrorAs` before we go GA, I'm fine leaving it for now.